### PR TITLE
Partial Tiles

### PR DESCRIPTION
**Pull request by adrianmroz-allegro**
Task removes refs on `FilterTiles` etc. and lifts state for partial tiles so `DimensionsMeasuresPanel` can use regular callbacks instead of blindly calling instance methods.

### CODE REVIEW

File: src/client/components/filter-tile/filter-tiles-row.tsx [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+      addPartialFilter(dimension, position);
+    }
+  }
+
+  private dropFilter(clause: FilterClause, position: DragPosition) {
+    const { clicker, essence: { filter } } = this.context;
+    const newFilter = position.isReplace()
+      ? filter.replaceByIndex(position.replace, clause)
+      : filter.insertByIndex(position.insert, clause);
+    !filter.equals(newFilter) && clicker.changeFilter(newFilter);
```
Comment (adrianmroz-allegro) [inline src/client/components/filter-tile/filter-tiles-row.tsx:160]: Ugh, to clever.

File: src/client/components/filter-menu/string-filter-menu/string-filter-menu.tsx [hunk 2]
```diff
@@ -60,19 +57,11 @@ export class StringFilterMenu extends React.Component<StringFilterMenuProps, Str
   onSelectFilterOption = (filterMode: FilterMode) => this.setState({ filterMode });
 
   updateFilter: (clause: FilterClause) => Filter = clause => {
```
Comment (adrianmroz-allegro) [inline src/client/components/filter-menu/string-filter-menu/string-filter-menu.tsx:59]: Should be unused.

File: src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx [hunk 3]
```diff
@@ -118,10 +119,11 @@ export class FixedTimeTab extends React.Component<FixedTimeTabProps, FixedTimeTa
   }
 
   isFilterDifferent(): boolean {
-    const { essence: { filter, timeShift } } = this.props;
+    const { essence: { filter, timeShift }, dimension } = this.props;
     const newTimeShift = this.constructTimeShift();
-    const newFilter = this.constructFixedFilter();
-    return !filter.equals(newFilter) || !timeShift.equals(newTimeShift);
+    const oldClause = filter.getClauseForDimension(dimension);
```
Comment (adrianmroz-allegro) [inline src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx:124]: We have a guarantee that oldClause exists, but this guarantee is not expressed in type system. Something to improve sometimes.

File: src/client/components/filter-tile/filter-tiles-row.tsx [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+  openFilterMenu = (clause: FilterClause) => this.setState({ openedClause: clause });
+
+  closeFilterMenu = () => this.setState({ openedClause: null });
+
+  openOverflowMenu = () => this.setState({ overflowOpen: true });
+
+  closeOverflowMenu = () => this.setState({ overflowOpen: false });
+
+  canDrop(): boolean {
```
Comment (adrianmroz-allegro) [inline src/client/components/filter-tile/filter-tiles-row.tsx:73]: A lot of the code here is very similar to SeriesTilesRow and SplitTilesRow but they differ even so slightly. I don't think we should push for DRY here.

File: src/client/components/filter-tile/filter-tiles-row.tsx [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    }
+  };
+
+  private dropDimension(dimension: Dimension, position: DragPosition) {
+    const { essence: { filter, dataCube } } = this.context;
+    const { addPartialFilter } = this.props;
+    let tryingToReplaceTime = false;
+    if (position.replace !== null) {
+      const targetClause = filter.clauses.get(position.replace);
+      tryingToReplaceTime = targetClause && targetClause.reference === dataCube.timeAttribute;
```
Comment (adrianmroz-allegro) [inline src/client/components/filter-tile/filter-tiles-row.tsx:148]: Maybe early return would be better?

File: src/client/components/series-tile/series-tiles-row.tsx [hunk 6]
```diff
@@ -60,22 +60,6 @@ export class SeriesTilesRow extends React.Component<SeriesTilesRowProps, SeriesT
     return menuStage && getMaxItems(menuStage.width, series.count());
   }
 
-  // This will be called externally
-  appendDirtySeries(series: Series) {
```
Comment (adrianmroz-allegro) [inline src/client/components/series-tile/series-tiles-row.tsx:64]: Yay!

File: src/client/components/series-tile/series-tiles-row.tsx [hunk 7]
```diff
@@ -193,34 +179,36 @@ export class SeriesTilesRow extends React.Component<SeriesTilesRowProps, SeriesT
   }
 
   appendMeasureSeries = (measure: Measure) => {
+    const { addPartialSeries } = this.props;
+    const { clicker, essence } = this.context;
     const series = fromMeasure(measure);
     const isMeasureSeries = series instanceof MeasureSeries;
     const isUniqueQuantile = !this.context.essence.series.hasSeries(series);
```
Comment (adrianmroz-allegro) [inline src/client/components/series-tile/series-tiles-row.tsx:186]: This checks only if it's unique. Add instanced QuantileSeries check also!

File: src/client/views/cube-view/partial-tiles-provider.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        series,
+        position
+      }
+    });
+  };
+
+  render() {
+    const { children } = this.props;
+    const { tile } = this.state;
+    const series = tile && tile.kind === "series" ? tile : null;
```
Comment (adrianmroz-allegro) [inline src/client/views/cube-view/partial-tiles-provider.ts:89]: We have single source of truth. It enforces that we have series or filter at the same time. But we still can provide easier API to consumers with distribution between values. And type systems checks if we assign correct sub-type to correct field.

File: src/client/components/filter-tile/filter-tiles-row.tsx (hunk 4 above)
Comment (fpoon) [reply src/client/components/filter-tile/filter-tiles-row.tsx:73]: If you suspect, that in the future this code would drift apart from SeriesTilesRow and SplitTilesRow I would recommend keeping it separate. However, if that's not the case, I would merge it with the rest

File: src/client/components/filter-tile/filter-tiles-row.tsx (hunk 5 above)
Comment (fpoon) [reply src/client/components/filter-tile/filter-tiles-row.tsx:148]: I agree

**Review by fpoon — APPROVED** [review]

**Pull request merged**